Support manual composed bringup for Nav2 applications (#2555)
ros-planning/navigation2: nav2_bringup/README.md

# nav2_bringup

The `nav2_bringup` package is an example bringup system for Nav2 applications. 

> This is a very flexible example for nav2 bring up that can be modified for different maps/robots/hardware/worlds/etc. It is our expectation for an application specific thing you're mirroring `nav2_bringup` package and modifying it for your specific  maps/robots/bringup needs. So this is an applied and working demonstration for the default system bringup with many options that can be easily modified. 

Reviewer: ```suggestion
This is a very flexible example for nav2 bringup that can be modified for different maps/robots/hardware/worlds/etc. It is our expectation for an application specific robot system that you're mirroring `nav2_bringup` package and modifying it for your specific maps/robots/bringup needs. This is an applied and working demonstration for the default system bringup with many options that can be easily modified. 
```

Usual robot stacks will have a `<robot_name>_nav` package with config/bringup files and this is that for the general case to base a specific robot system off of.

> Composed bringup (based on  [ROS2 Composition](https://docs.ros.org/en/galactic/Tutorials/Composition.html) ) is optional for user, in other word, compose all Nav2 nodes in a single process instead of launching these nodes separately, which is useful for embedded systems users that need to make optimizations due to harsh resource constraints.  

Reviewer: ```suggestion
Composed bringup (based on  [ROS2 Composition](https://docs.ros.org/en/galactic/Tutorials/Composition.html)) is optional for users. It can be used to compose all Nav2 nodes in a single process instead of launching these nodes separately, which is useful for embedded systems users that need to make optimizations due to harsh resource constraints.  
```

> * some discussions about performance improvement of composed bringup could be found here: https://discourse.ros.org/t/nav2-composition/22175.

Reviewer: ```suggestion
* Some discussions about performance improvement of composed bringup could be found here: https://discourse.ros.org/t/nav2-composition/22175.
```

> * currently, manual composition is used in this package,  and dynamic composition is more flexible than manual composition, but it could be applied in nav2 due to various issues, you could find more details here: https://github.com/ros-planning/navigation2/issues/2147.

Reviewer: ```suggestion
* Currently, manual composition is used in this package. Dynamic composition is more flexible than manual composition, but is not currently applied in nav2 due to various issues, you could find more details here: https://github.com/ros-planning/navigation2/issues/2147.
```

> ### Terminal 3: Launch Nav2

Reviewer: At the top of this readme adding that this is a very flexible example for nav2 bring up that can be modified for different maps/robots/hardware/worlds/etc. It is our expectation for an application specific thing you're mirroring nav2_bringup package and modifying it for your sp. maps/robots/bringup needs. So this is an applied and working demonstration for the default system bringup with many options that can be easily modified. 

Usual robot stacks will have a `<robot_name>_nav` package with config/bringup files and this is that for the general case to base a specific robot system off of

Reviewer: Also, some discussion about why we do manual composition rather than dynamic composition in the launch file due to various issues in place (parameters, lifecycle components not playing well, etc). Eventually we'd do dynamic components but for now, manually composed. Link the relevent tickets / issues so that if they get resolved someone can take note and potentially enable dynamic composition

normal bringup

```bash
source /opt/ros/dashing/setup.bash
ros2 launch nav2_bringup bringup_launch.py use_sim_time:=True autostart:=True \
map:=<full/path/to/map.yaml>
```

manual composed bringup

```bash
source /opt/ros/dashing/setup.bash
ros2 launch nav2_bringup manual_composed_bringup_launch.py use_sim_time:=True autostart:=True \
map:=<full/path/to/map.yaml>
```

* composed bringup is based on [ROS2 Composition](https://docs.ros.org/en/galactic/Tutorials/Composition.html), which is useful for embedded systems users that need to make optimizations due to harsh resource constraints.
* some discussions about performance improvement could be found at https://discourse.ros.org/t/nav2-composition/22175

> 

Reviewer: Unrelated to this line: but you said you registered these nodes as components to do some testing? Did that work out OK? If so, please submit a new PR with the composition registration so that once we can dynamically compose the nodes, all that stuff is already ready to go.

Author: > you said you registered these nodes as components to do some testing?

yes,  these nodes could registered as components.

> Did that work out OK?

no, it can't work as expected, because internal node in components couldn't load parameters (e.g. `local_costmap` in `controller_server`),  you could find details here https://github.com/ros-planning/navigation2/issues/2147#issuecomment-915890304.

### Terminal 3: Launch Nav2

normal bringup

```bash
source /opt/ros/dashing/setup.bash
ros2 launch nav2_bringup bringup_launch.py use_sim_time:=True autostart:=True \
map:=<full/path/to/map.yaml>
```

manually composed bringup

```bash
source /opt/ros/dashing/setup.bash

> ros2 launch nav2_bringup composed_bringup_launch.py use_sim_time:=True autostart:=True \

Reviewer: I think this is now a commandline argument change, not a new launch file now. 

Non-composed bringup instructions need to set it to false. 

Also, capitalize "Manually Composed Bringup" and "Normal Bringup"
